bug(table): Conflict detection is silently bypassed when the writer starts from an empty table

### Apache Iceberg version

main (development)

### Please describe the bug 🐞

When a writer loads a table that has **no snapshot yet** on the target branch (empty table, or a branch that the writer is about to create), the `newConflictContext` helper returns an **empty** `concurrent` list — it treats the situation as "no concurrent activity". If another writer then commits a snapshot between the first writer's load time and its commit time, the conflict validators never see those concurrent commits. A `RowDelta` carrying equality-delete files will be accepted even though it should be rejected under `SERIALIZABLE` isolation.

## Affected component

`table/conflict_validation.go` — `newConflictContext()`

## Steps to reproduce

The test is self-contained and lives in **package `table`** (internal test) so it can call the unexported `newConflictContext` directly.  The helpers it uses (`newConflictTestMetadata`, `MainBranch`) are already in `table/conflict_validation_test.go` on `main`.

Save the file below as `table/bug_repro_empty_table_conflict_test.go` in an unmodified checkout of `main` and run:

```
go test ./table/ -run TestBugRepro_EmptyTableConflictDetectionBypassed -v
```

Expected output on unfixed upstream:

```
--- FAIL: TestBugRepro_EmptyTableConflictDetectionBypassed
    bug_repro_empty_table_conflict_test.go:76:
        Error:    "[]" should have 1 item(s), but has 0
        Messages: snapshot 42 must appear in concurrent when base has no branch head
```

Full source of `table/bug_repro_empty_table_conflict_test.go`:

```go
package table

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestBugRepro_EmptyTableConflictDetectionBypassed(t *testing.T) {
	// Writer loaded the table before any snapshot was committed on main.
	base := newConflictTestMetadata(t, nil)

	// Meanwhile another writer committed snapshot 42.
	head := int64(42)
	current := newConflictTestMetadata(t, &head)

	ctx, err := newConflictContext(base, current, MainBranch, nil, true)
	require.NoError(t, err)

	// Snapshot 42 was committed while Writer had no view of the branch.
	// It MUST appear in ctx.concurrent so that conflict validators (e.g.
	// RowDelta's serializable check) can inspect the data files it added.
	//
	// BUG: on upstream, ctx.concurrent is empty — snapshot 42 is invisible.
	// Any equality-delete commit from the empty-base writer will be accepted
	// without validating against the concurrent data.
	assert.Len(t, ctx.concurrent, 1,
		"snapshot 42 must appear in concurrent when base has no branch head")
}
```

## Root cause

The relevant section of `table/conflict_validation.go` on `main`:

```go
baseHead := base.SnapshotByName(branch)
if baseHead == nil {
    // Writer has no view of this branch yet (e.g. creating it) —
    // by definition there are no concurrent commits to validate against.
    return &conflictContext{current: current, branch: branch, fs: fs, caseSensitive: caseSensitive}, nil
}
```

When `baseHead` is `nil` the code returns immediately with an empty `concurrent` list, reasoning that there was "nothing to compare". But a nil `baseHead` simply means the writer did not see any snapshot when it loaded the table. Snapshots committed *after* that load but *before* this commit are still real concurrent activity that validators must examine.

## Concrete scenario where this matters

1. Table is empty. Worker A begins building an equality-delete commit (e.g. a streaming DELETE job that marks rows as deleted by field value).
2. Worker B appends new data files to the same partition and commits first.
   The table now has snapshot B.
3. Worker A calls `Commit`. Under `SERIALIZABLE` isolation, Worker A's equality-delete must be rejected: Worker B added data that Worker A's delete predicate may not cover — Worker A must reload and rebuild.
4. **Bug:** `newConflictContext` sees `baseHead == nil` and returns zero concurrent snapshots. The `RowDelta` validator never checks Worker B's data files. Worker A's commit is accepted and the equality deletes are applied against data that the writer never saw.

## Expected behaviour

When a writer started from an empty branch (no prior snapshot), all snapshots currently on that branch were committed after the writer loaded the table and are therefore concurrent. The returned `conflictContext.concurrent` list must contain all of them so validators can decide whether a conflict exists.

--- a/table/conflict_validation.go
+++ b/table/conflict_validation.go
@@ -192,10 +192,26 @@ func newConflictContext(base, current Metadata, branch string, fs iceio.IO, case
 
 	baseHead := base.SnapshotByName(branch)
 	if baseHead == nil {
-		// Writer has no view of this branch yet (e.g. creating it) —
-		// by definition there are no concurrent commits to validate
-		// against.
-		return &conflictContext{current: current, branch: branch, fs: fs, caseSensitive: caseSensitive}, nil
+		// Writer's base has no view of this branch (empty table or
+		// branch created concurrently). Every snapshot now reachable
+		// from the branch head is concurrent relative to base.
+		// Truncation (missing intermediate, cycle) is treated as
+		// divergent for the same reason AncestorsBetween's
+		// baseFound=false is: an under-counted concurrent list silently
+		// hides conflicts.
+		concurrent, complete := AncestorsOfChecked(currentHead.SnapshotID, current.SnapshotByID)
+		if !complete {
+			return nil, fmt.Errorf("%w: ancestry walk from %d on %s truncated; no base snapshot to bound against",
+				ErrCommitDiverged, currentHead.SnapshotID, branch)
+		}
+
+		return &conflictContext{
+			current:       current,
+			branch:        branch,
+			fs:            fs,
+			caseSensitive: caseSensitive,
+			concurrent:    concurrent,
+		}, nil
 	}
 
 	concurrent, baseFound := AncestorsBetween(currentHead.SnapshotID, baseHead.SnapshotID, current.SnapshotByID)

--- a/table/conflict_validation_test.go
+++ b/table/conflict_validation_test.go
@@ -145,6 +145,43 @@ func newConflictTestMetadataWithProps(t *testing.T, branchHead *int64, extraProp
 	return out
 }
 
+// newConflictTestMetadataWithChain builds metadata whose branch head
+// reaches back through ids[0]→ids[1]→...→ids[n-1] via ParentSnapshotID.
+// ids[len-1] becomes the branch head; ids[0] is the chain root.
+func newConflictTestMetadataWithChain(t *testing.T, ids []int64) Metadata {
+	t.Helper()
+	require.NotEmpty(t, ids)
+
+	schema := iceberg.NewSchema(0,
+		iceberg.NestedField{ID: 1, Name: "id", Type: iceberg.PrimitiveTypes.Int64, Required: true},
+	)
+	props := iceberg.Properties{PropertyFormatVersion: "2"}
+	meta, err := NewMetadata(schema, iceberg.UnpartitionedSpec, UnsortedSortOrder, "file:///tmp/conflict-test", props)
+	require.NoError(t, err)
+
+	builder, err := MetadataBuilderFromBase(meta, "")
+	require.NoError(t, err)
+
+	for i, id := range ids {
+		snap := Snapshot{
+			SnapshotID:     id,
+			SequenceNumber: int64(i + 1),
+			TimestampMs:    meta.LastUpdatedMillis() + int64(i+1),
+			Summary:        &Summary{Operation: OpAppend},
+		}
+		if i > 0 {
+			parent := ids[i-1]
+			snap.ParentSnapshotID = &parent
+		}
+		require.NoError(t, builder.AddSnapshot(&snap))
+	}
+	require.NoError(t, builder.SetSnapshotRef(MainBranch, ids[len(ids)-1], BranchRef))
+	out, err := builder.Build()
+	require.NoError(t, err)
+
+	return out
+}
+
 func TestNewConflictContext_NoConcurrentCommits(t *testing.T) {
 	// base and current point at the same snapshot → zero concurrent
 	// snapshots, no error.
@@ -157,15 +194,35 @@ func TestNewConflictContext_NoConcurrentCommits(t *testing.T) {
 }
 
 func TestNewConflictContext_WriterHasNoBranchView(t *testing.T) {
-	// Writer is creating the branch for the first time (base has no
-	// ref yet) — there is nothing concurrent to validate against.
+	// Writer loaded the table before any snapshot was committed on
+	// the branch (empty base). A snapshot then appeared on the
+	// branch — it must be reported as concurrent so validators that
+	// care about concurrent writes (e.g. RowDelta under
+	// SERIALIZABLE) can inspect its data files.
 	base := newConflictTestMetadata(t, nil)
 	head := int64(7)
 	current := newConflictTestMetadata(t, &head)
 
 	ctx, err := newConflictContext(base, current, MainBranch, nil, true)
 	require.NoError(t, err)
-	assert.Empty(t, ctx.concurrent)
+	require.Len(t, ctx.concurrent, 1)
+	assert.Equal(t, int64(7), ctx.concurrent[0].SnapshotID)
+}
+
+func TestNewConflictContext_EmptyBaseEnumeratesFullAncestry(t *testing.T) {
+	// Empty base, current branch head reaches back through a chain of
+	// three snapshots — all three must surface as concurrent in
+	// reverse-chronological order so validators see the full set the
+	// writer never observed.
+	base := newConflictTestMetadata(t, nil)
+	current := newConflictTestMetadataWithChain(t, []int64{10, 11, 12})
+
+	ctx, err := newConflictContext(base, current, MainBranch, nil, true)
+	require.NoError(t, err)
+	require.Len(t, ctx.concurrent, 3)
+	assert.Equal(t, int64(12), ctx.concurrent[0].SnapshotID)
+	assert.Equal(t, int64(11), ctx.concurrent[1].SnapshotID)
+	assert.Equal(t, int64(10), ctx.concurrent[2].SnapshotID)
 }
 
 func TestNewConflictContext_MissingCurrentBranch(t *testing.T) {

--- a/table/snapshot_ancestry.go
+++ b/table/snapshot_ancestry.go
@@ -33,34 +33,51 @@ type SnapshotLookup func(id int64) *Snapshot
 // The returned slice may be truncated if an intermediate snapshot is
 // missing from the lookup (e.g. expired) or if a cycle is encountered.
 // Callers that need to distinguish a complete walk from a truncated one
-// should use IsAncestorOf against the expected oldest ancestor.
+// should use AncestorsOfChecked instead.
 func AncestorsOf(snapshotID int64, lookup SnapshotLookup) []Snapshot {
+	ancestors, _ := AncestorsOfChecked(snapshotID, lookup)
+
+	return ancestors
+}
+
+// AncestorsOfChecked is AncestorsOf with completeness tracking. The
+// second return value is true when the walk terminated at a snapshot
+// with no parent (a clean root). It is false when the walk was
+// truncated by an unresolvable starting snapshot, a missing intermediate
+// snapshot, or a cycle in malformed metadata.
+//
+// Callers performing conflict detection (where a truncated ancestry
+// equates to under-counting concurrent snapshots) MUST treat
+// complete=false as divergent and refuse the commit, mirroring
+// AncestorsBetween's baseFound=false contract. When complete is false
+// the returned slice is the partial walk before truncation —
+// diagnostic context only, NOT an enumerable ancestry.
+//
+// Snapshots are returned by value in reverse-chronological order.
+// Returns an empty slice and false when snapshotID cannot be resolved.
+func AncestorsOfChecked(snapshotID int64, lookup SnapshotLookup) ([]Snapshot, bool) {
 	var ancestors []Snapshot
 	visited := make(map[int64]struct{})
 
 	id := snapshotID
 	for {
 		if _, seen := visited[id]; seen {
-			// Cycle detection — stop walking. A well-formed metadata will
-			// never hit this, but defensive guard avoids infinite loops.
-			break
+			return ancestors, false
 		}
 		visited[id] = struct{}{}
 
 		snap := lookup(id)
 		if snap == nil {
-			break
+			return ancestors, false
 		}
 
 		ancestors = append(ancestors, *snap)
 
 		if snap.ParentSnapshotID == nil {
-			break
+			return ancestors, true
 		}
 		id = *snap.ParentSnapshotID
 	}
-
-	return ancestors
 }
 
 // AncestorsBetween returns the snapshots from latestID (inclusive) down

--- a/table/snapshot_ancestry_test.go
+++ b/table/snapshot_ancestry_test.go
@@ -101,6 +101,45 @@ func TestAncestorsOf_CycleDefense(t *testing.T) {
 	assert.Equal(t, []int64{1, 2}, snapshotIDs(got))
 }
 
+func TestAncestorsOfChecked_LinearChain(t *testing.T) {
+	got, complete := AncestorsOfChecked(5, buildChain(5))
+
+	assert.True(t, complete, "walk reached the root")
+	assert.Equal(t, []int64{5, 4, 3, 2, 1}, snapshotIDs(got))
+}
+
+func TestAncestorsOfChecked_Unknown(t *testing.T) {
+	got, complete := AncestorsOfChecked(99, buildChain(3))
+
+	assert.False(t, complete, "unresolvable starting snapshot is not a complete walk")
+	assert.Empty(t, got)
+}
+
+func TestAncestorsOfChecked_BrokenChain(t *testing.T) {
+	// Chain 1 ← 2 ← 3 with snapshot 2 missing — the walk stops at 3 and
+	// the missing parent must be reported as truncation.
+	chain := buildChain(3)
+	lookup := func(id int64) *Snapshot {
+		if id == 2 {
+			return nil
+		}
+
+		return chain(id)
+	}
+
+	got, complete := AncestorsOfChecked(3, lookup)
+
+	assert.False(t, complete, "missing intermediate must surface as truncation")
+	assert.Equal(t, []int64{3}, snapshotIDs(got))
+}
+
+func TestAncestorsOfChecked_CycleDefense(t *testing.T) {
+	got, complete := AncestorsOfChecked(1, buildCycle())
+
+	assert.False(t, complete, "cycle must surface as truncation")
+	assert.Equal(t, []int64{1, 2}, snapshotIDs(got))
+}
+
 func TestAncestorsBetween_LinearChain(t *testing.T) {
 	lookup := buildChain(5)
 

